### PR TITLE
create measurement class and io function

### PR DESCRIPTION
**Pull request by roussel-ryan**
Adds abstract classes for file io, measurements

### CODE REVIEW

**Comment by roussel-ryan** [comment]
lcls_tools_objects

**Comment by roussel-ryan** [comment]
An example of how a basic image diagnostic would work:
```
class ScreenBeamProfile(Measurement):
    name = "beam_profile"
    device: Screen
    image_processer: Any
    fitting_tool: FittingTool
    fit_profile: bool = True

    n_shots: int = 5
    charge_window: Optional[ChargeWindow] = None

    def measure(self) -> dict:
        images = []
        while len(images) < self.n_shots:
            measurement = self.single_measure()
            if len(measurement):
                images += [self.single_measure()]

        # fit profile if requested
        if self.fit_profile:
            results = []
            for ele in images:
                results += [self.fitting_tool.fit_image(ele["processed_image"])]

            # combine images with profile info
            final_results = {}
        else:
            # concat list elements into a single dict
            final_results = {}

        if self.save_data:
            self.dump_controller.dump_data_to_file(final_results, self)

        return final_results

    def single_measure(self) -> dict:
        # measure profiles
        # get raw data
        raw_image = self.device.image

        # get ICT measurements and return None if not in window
        if self.charge_window is not None:
            if not self.charge_window.in_window():
                return {}

        processed_image = self.image_processer.process(raw_image)
        return {"raw_image": raw_image, "processed_image": processed_image}
```

File: lcls_tools/common/devices/ict.py
Comment (nneveu) [inline lcls_tools/common/devices/ict.py:9]: do you mean ICT here instead of screen?

File: lcls_tools/common/devices/ict.py
Comment (nneveu) [inline lcls_tools/common/devices/ict.py:8]: we would need to update the reader script to pull ICT in yaml files too. Do you have a target ICT/name I could add?

File: lcls_tools/common/measurements/measurement.py
Comment (nneveu) [inline lcls_tools/common/measurements/measurement.py:11]: should we have metadata placeholders here that all measurements should have? Area? device locations? rmats?

File: lcls_tools/common/measurements/measurement.py
Comment (nneveu) [inline lcls_tools/common/measurements/measurement.py:15]: Trying to think of what else besides metadata would be common to all measurements? I'm sure there's something.

File: lcls_tools/common/io.py
Comment (nneveu) [inline lcls_tools/common/io.py:7]: small preference, but I think I would go for H5reader and H5writer?

File: lcls_tools/common/measurements/measurement.py
Comment (nneveu) [reply lcls_tools/common/measurements/measurement.py:11]: Maybe beam path over area...

File: lcls_tools/common/measurements/beam_charge.py
Comment (nneveu) [inline lcls_tools/common/measurements/beam_charge.py:6]: I would go for BeamCharge vs. BeamChargeMeasurement?

File: lcls_tools/common/measurements/beam_charge.py
Comment (nneveu) [inline lcls_tools/common/measurements/beam_charge.py:11]: Assign this to a variable, as we'll prob add more things to the dict that gets returned?

**Review by nneveu — COMMENTED** [review]
Looks like a great start, thanks Ryan!

File: lcls_tools/common/devices/ict.py
Comment (roussel-ryan) [reply lcls_tools/common/devices/ict.py:8]: no idea

File: lcls_tools/common/measurements/measurement.py
Comment (roussel-ryan) [reply lcls_tools/common/measurements/measurement.py:11]: the measurements will contain device objects, which themselves contain the metadata. We can add a property to the measurement object that collects this info for a user if necessary

File: lcls_tools/common/measurements/measurement.py
Comment (roussel-ryan) [reply lcls_tools/common/measurements/measurement.py:15]: I imagine something will come up as we implement more measurements

File: lcls_tools/common/measurements/beam_charge.py
Comment (roussel-ryan) [reply lcls_tools/common/measurements/beam_charge.py:11]: sure thats fine, but should be done in a device specific PR. This is just a simple placeholder

**Comment by roussel-ryan** [comment]
Keep in mind that the scope of this PR is specific to constructing the overall framework (abstract classes) and changes can be easily made to the implementations of specific functions. Any implementations shown here are just examples to understand how the classes interact with one another

File: lcls_tools/common/measurements/beam_charge.py
Comment (nneveu) [reply lcls_tools/common/measurements/beam_charge.py:11]: If it's something we know we will do, I lean towards just adding it in now for style/format guidance to future contributors. At some point it will be hard to keep track of things.

**Comment by roussel-ryan** [comment]
> Let's use `NotImplemented` instead of pass. I think it's fine to have this as a framework, but if there are things we know need to add, I'd prefer to take the extra time to do it now. That way it stands as a better example. I can commit some of the changes.

Ok thats fine if you want to mostly finish the ICT measurement example, I'm holding off on adding device specific code (ie. screen code, @phys-cgarnier's projection fitting code) until this PR is merged.

**Comment by roussel-ryan** [comment]
I decided against specifying a separate multi-shot wrapper class since each type of measurement might have different ways of effectively measuring and calculating statistics

**Review by nneveu — APPROVED** [review]

**Comment by nneveu** [comment]
Note, will add an issue to update yaml files to have a real ICT example.

**Pull request merged**
